src/posts/2022-07-13-L4LB-XDP/index.md

```diff
@@ -0,0 +1,225 @@
+--bpf-lb-algorithm=maglev \
+--bpf-lb-mode=dsr \
+--bpf-lb-acceleration=native \
+--bpf-lb-dsr-dispatch=ipip \
+--devices=eth0 \
+--datapath-mode=lb-only \
+--enable-l7-proxy=false \
+--tunnel=disabled \
+--install-iptables-rules=false \
+--enable-bandwidth-manager=false \
+--enable-local-redirect-policy=false \
+--enable-hubble=false \
+--enable-l7-proxy=false \
+--preallocate-bpf-maps=false \
+--disable-envoy-version-check=true \
+-auto-direct-node-routes=false \
+--enable-ipv4=true \
+--enable-ipv6=true
+```
+
+We serve around 3k services and use 30+ L7 nodes so we quickly reached the defaults for lbmap size. However, it is possible to extend it so we added the `--bpf-lb-map-max 512000` option.
+
+# Setting Up Services
+Cilium provides an API to set up the lbmaps. We used the following command to configure all the services:
+ 
+```
+cilium service update --id $idx --frontend "$svc" --backends "$backends" --k8s-node-port
+```
+ 
+The `frontend` stands for each VIP service and `backends` for the L7 nodes (e.g. `cilium service update --id 1 --frontend "10.248.11.13:7047" --backends "10.246.3.34:7047,10.246.39.33:7047,10.246.39.34:7047" --k8s-node-port`).
+ 
+For BGP announcement, [we use BIRD](https://docs.cilium.io/en/stable/gettingstarted/bird/) so this part is fairly straightforward:
+```
+# systemctl start bird
+# systemctl start bird6
+```
+
+# Comparison Under Load
+* [SynFlood](https://www.cloudflare.com/learning/ddos/syn-flood-ddos-attack/)
+* [MoonGen](https://github.com/emmericp/MoonGen)
+* 1 CPU for generating traffic
+* 64B packets
+ 
+At first, we decided to compare IPVS with L4LB under synthetic test/load generated by a single client running a MoonGen with 1 CPU and small 64B packets with TCP SYN set. Packets (tcp segments) sent out by the MoonGen generator were configured to randomize source IP address and TCP source port so that the traffic is distributed across all receive rx queues as our NIC cards are configured to use 4 tuple (src IP, dst IP, src TCP port, dst TCP port) hashing:
+
+```
+# ethtool -n eth0 rx-flow-hash tcp4
+TCP over IPV4 flows use these fields for computing Hash flow key:
+IP SA
+IP DA
+L4 bytes 0 & 1 [TCP/UDP src port]
+L4 bytes 2 & 3 [TCP/UDP dst port]
+```
+During the test run we used another client running a simple GET request (using curl in a while loop) to see how the server was able to handle the requests.
+ 
+The test setup looked like this:
+
+![Seznam.cz testing](testing.svg)
+ 
+## Results
+ 
+In both scenarios (scenario #1 IPVS and scenario #2 L4LB) the MoonGen client was configured to generate 1Mpps (million packets per second) and 3Mpps.
+Each output screenshot below is taken from the corresponding server - either the server IPVS/L4LB under the test and the `curl` client. For L4LB XDP, both 1Mpps and 3Mpps were too easy and did not have any performance impact. We instead started with 10Mpps and only saw an impact at 14.8 Mpps which was probably due to constraints of the NIC rather than L4LB.
```

Comment:
```suggestion
Each output screenshot below is taken from the corresponding server - either the server IPVS/L4LB under test or the `curl` client. For L4LB XDP, both 1Mpps and 3Mpps were too easy and did not have any performance impact. We instead started with 10Mpps and only saw an impact at 14.8 Mpps which was probably due to constraints of the NIC rather than L4LB.
```